<?xml version='1.0'?>
<robot name="legged_robot" xmlns:xacro="http://www.ros.org/wiki/xacro">
  <!--static>false</static--> <!-- true is static model-->

<!-- Included URDF/XACRO Files -->
<xacro:include filename="$(find wamot)/urdf/materials.urdf.xacro" />

<!--configuration-->
<property name="base_mass" value="10.0" />
<property name="base_x_size" value="0.5" />
<property name="base_y_size" value="0.3" />
<property name="base_z_size" value="0.05" />
<property name="leg_base_height" value="0.050" />
<property name="leg_base_radius" value="0.025" />
<property name="lower_leg_height" value="0.3" />
<property name="lower_leg_radius" value="0.02" />
<property name="lower_leg_mass" value="2.0" />
<property name="upper_leg_height" value="0.3" />
<property name="upper_leg_radius" value="0.02" />
<property name="upper_leg_mass" value="2.0" />

<!--macro-->
<macro name="box_inertia" params="m x y z">
<inertia  ixx="${m*(y*y+z*z)/12.}" ixy = "0." ixz = "0."
iyy="${m*(x*x+z*z)/12.}" iyz = "0."
izz="${m*(x*x+z*z)/12.}" />
</macro>

<macro name="cylinder_inertia" params="m r h">
<inertia  ixx="${m*(3*r*r+h*h)/12}" ixy = "0" ixz = "0"
iyy="${m*(3*r*r+h*h)/12}" iyz = "0"
izz="${m*r*r/2}" />
</macro>

<link name="base_footprint"/>

<gazebo reference="base_footprint">
<turnGravityOff>false</turnGravityOff>
</gazebo>

<joint name="base_link_joint" type="fixed">
<origin xyz="0 0 0" rpy="0 0 0" />
<parent link="base_footprint"/>
<child link="torso" />
</joint>

  <!--body-->
  <link name="torso">
    <inertial>
    <origin xyz="0 0 0" rpy="0 0 0" />
    <mass value="${base_mass}" />
    <box_inertia  m="${base_mass}" x="${base_x_size}" y="${base_y_size}" z="${base_z_size}"/>
    </inertial>
    <collision name='collision'>
    <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
	<box size="${base_x_size} ${base_y_size} ${base_z_size}"/>
      </geometry>
    </collision>
    <visual name='visual'>
    <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
	<box size="${base_x_size} ${base_y_size} ${base_z_size}"/>
      </geometry>
      <material name="White" />
    <material name="White" />
   </visual>
  </link>

  <link name='base_left_fore'>
  <visual name='visual'>
    <geometry>
      <cylinder length="${leg_base_height}" radius="${leg_base_radius}"/>
    </geometry>
  <material name="White" />
   </visual>
  </link>

  <link name='base_left_rear'>
  <visual name='visual'>
    <geometry>
      <cylinder length="${leg_base_height}" radius="${leg_base_radius}"/>
    </geometry>
  <material name="White" />
   </visual>
  </link>
  
  <link name='base_right_fore'>
  <visual name='visual'>
    <geometry>
      <cylinder length="${leg_base_height}" radius="${leg_base_radius}"/>
    </geometry>
  <material name="White" />
   </visual>
  </link>
 
  <link name='base_right_rear'>
    <visual name='visual'>
      <geometry>
	<cylinder length="${leg_base_height}" radius="${leg_base_radius}"/>
      </geometry>
    <material name="White" />
   </visual>
  </link>

  <link name='link_left_fore_lower'>
    <!--self_collide>0</self_collide-->
    <inertial>
     <origin xyz="0 0 0.15" rpy="0 0 0" />
      <mass value="${lower_leg_mass}" />
      <cylinder_inertia  m="${lower_leg_mass}" r="${lower_leg_radius}" h="${lower_leg_height}" />
    </inertial>
    <collision name='collision'>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <geometry>
	<cylinder length="${lower_leg_height}" radius="${lower_leg_radius}"/>
      </geometry>
      <!--surface>
          <friction>
            <ode>
              <mu>100000.0</mu>
	      <mu2>100000.0</mu2>
              <slip1>0.0</slip1>
              <slip2>0.0</slip2>
            </ode>
          </friction>
      </surface-->
    </collision>
    <visual name='visual'>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <geometry>
	<cylinder length="${lower_leg_height}" radius="${lower_leg_radius}"/>
      </geometry>
    <material name="White" />
   </visual>
  </link>

  <link name='link_left_rear_lower'>
    <!--self_collide>0</self_collide-->
    <inertial>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <mass value="${lower_leg_mass}" />
      <cylinder_inertia  m="${lower_leg_mass}" r="${lower_leg_radius}" h="${lower_leg_height}" />
    </inertial>
    <collision name='collision'>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <geometry>
	<cylinder length="${lower_leg_height}" radius="${lower_leg_radius}"/>
      </geometry>
      <!--surface>
          <friction>
            <ode>
              <mu>100000.0</mu>
              <mu2>100000.0</mu2>
              <slip1>0.0</slip1>
              <slip2>0.0</slip2>
            </ode>
          </friction>
      </surface-->
    </collision>
    <visual name='visual'>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <geometry>
	<cylinder length="${lower_leg_height}" radius="${lower_leg_radius}"/>
      </geometry>
    <material name="White" />
   </visual>
  </link>
  
  <link name='link_right_fore_lower'>
  <!--self_collide>0</self_collide-->
    <inertial>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <mass value="${lower_leg_mass}" />
      <cylinder_inertia  m="${lower_leg_mass}" r="${lower_leg_radius}" h="${lower_leg_height}" />
    </inertial>
    <collision name='collision'>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <geometry>
	<cylinder length="${lower_leg_height}" radius="${lower_leg_radius}"/>
      </geometry>
      <!--surface>
          <friction>
            <ode>
              <mu>100000.0</mu>
              <mu2>100000.0</mu2>
              <slip1>0.0</slip1>
              <slip2>0.0</slip2>
            </ode>
          </friction>
      </surface-->
    </collision>
    <visual name='visual'>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <geometry>
	<cylinder length="${lower_leg_height}" radius="${lower_leg_radius}"/>
      </geometry>
    <material name="White" />
   </visual>
  </link>
  
  <link name='link_right_rear_lower'>
  <!--self_collide>0</self_collide-->
    <inertial>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <mass value="${lower_leg_mass}" />
      <cylinder_inertia  m="${lower_leg_mass}" r="${lower_leg_radius}" h="${lower_leg_height}" />
    </inertial>
    <collision name='collision'>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <geometry>
	<cylinder length="${lower_leg_height}" radius="${lower_leg_radius}"/>
      </geometry>
      <!--surface>
          <friction>
            <ode>
              <mu>100000.0</mu>
              <mu2>100000.0</mu2>
              <slip1>0.0</slip1>
              <slip2>0.0</slip2>
            </ode>
          </friction>
      </surface-->
    </collision>
    <visual name='visual'>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <geometry>
	<cylinder length="${lower_leg_height}" radius="${lower_leg_radius}"/>
      </geometry>
    <material name="White" />
   </visual>
  </link>
  
  <link name='link_left_fore_upper'>
  <!--self_collide>0</self_collide-->
    <inertial>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <mass value="${upper_leg_mass}" />
      <cylinder_inertia  m="${upper_leg_mass}" r="${upper_leg_radius}" h="${upper_leg_height}" />
    </inertial>
    <collision name='collision'>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <geometry>
	<cylinder length="${upper_leg_height}" radius="${upper_leg_radius}"/>
      </geometry>
    </collision>
    <visual name='visual'>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <geometry>
	<cylinder length="${upper_leg_height}" radius="${upper_leg_radius}"/>
      </geometry>
    <material name="White" />
   </visual>
  </link>

  <link name='link_left_rear_upper'>
  <!--self_collide>0</self_collide-->
    <inertial>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <mass value="${upper_leg_mass}" />
      <cylinder_inertia  m="${upper_leg_mass}" r="${upper_leg_radius}" h="${upper_leg_height}" />
    </inertial>
    <collision name='collision'>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <geometry>
	<cylinder length="${upper_leg_height}" radius="${upper_leg_radius}"/>
      </geometry>
    </collision>
    <visual name='visual'>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <geometry>
	<cylinder length="${upper_leg_height}" radius="${upper_leg_radius}"/>
      </geometry>
    <material name="White" />
   </visual>
  </link>
  
  <link name='link_right_fore_upper'>
  <!--self_collide>0</self_collide-->
    <inertial>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <mass value="${upper_leg_mass}" />
      <cylinder_inertia  m="${upper_leg_mass}" r="${upper_leg_radius}" h="${upper_leg_height}" />
    </inertial>
    <collision name='collision'>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <geometry>
	<cylinder length="${upper_leg_height}" radius="${upper_leg_radius}"/>
      </geometry>
    </collision>
    <visual name='visual'>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <geometry>
	<cylinder length="${upper_leg_height}" radius="${upper_leg_radius}"/>
      </geometry>
    <material name="White" />
   </visual>
  </link>

  <link name='link_right_rear_upper'>
  <!--self_collide>0</self_collide-->
    <origin xyz="0 0 0.15" rpy="0 0 0" />
    <inertial>
      <mass value="${upper_leg_mass}" />
      <cylinder_inertia  m="${upper_leg_mass}" r="${upper_leg_radius}" h="${upper_leg_height}" />
    </inertial>
    <collision name='collision'>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <geometry>
	<cylinder length="${upper_leg_height}" radius="${upper_leg_radius}"/>
      </geometry>
    </collision>
    <visual name='visual'>
      <origin xyz="0 0 0.15" rpy="0 0 0" />
      <geometry>
	<cylinder length="${upper_leg_height}" radius="${upper_leg_radius}"/>
      </geometry>
    <material name="White" />
   </visual>
  </link>


  <!--joint-->
  <!--left fore 1st joint-->
  <joint type="revolute" name="hinge_lf1"> 
    <parent link="torso"/>
    <child link="base_left_fore"/>
    <origin xyz="0.23 0.17 0.0" rpy="1.5707 0 0" />
    <axis xyz="0 0 1" rpy="0 0 0"/>
    <limit effort="30" velocity="1.0" lower="-2.2" upper="0.7" />
  </joint>

  <!--left rear 1st joint-->
  <joint type="revolute" name="hinge_lr1">  
    <parent link="torso"/>
    <child link="base_left_rear"/>
    <origin xyz="-0.23 0.17 0.0" rpy="1.5707 0 0" />
    <axis xyz="0 0 1" rpy="0 0 0"/>
    <limit effort="30" velocity="1.0" lower="-2.2" upper="0.7" />
  </joint>
  
  <!--right fore 1st joint-->
  <joint type="revolute" name="hinge_rf1"> 
    <parent link="torso"/>
    <child link="base_right_fore"/>
    <origin xyz="0.23 -0.17 0.0" rpy="1.5707 0 0" />
    <axis xyz="0 0 1" rpy="0 0 0"/>
    <limit effort="30" velocity="1.0" lower="-2.2" upper="0.7" />
  </joint>

  <!--right rear 1st joint--> 
  <joint type="revolute" name="hinge_rr1"> 
    <parent link="torso"/>  
    <child link="base_right_rear"/>
    <origin xyz="-0.23 -0.17 0.0" rpy="1.5707 0 0" />
    <axis xyz="0 0 1" rpy="0 0 0"/>
    <limit effort="30" velocity="1.0" lower="-2.2" upper="0.7" />
  </joint>
  
  <!--left fore 2nd joint--> 
  <joint type="revolute" name="hinge_lf2">
    <parent link="base_left_fore"/>  
    <child link="link_left_fore_upper"/>
    <origin xyz="0 0 0" rpy="1.5707 0 0" />
    <axis xyz="0 1 0" rpy="0 0 0"/>
    <limit effort="30" velocity="1.0" lower="-2.2" upper="0.7" />
  </joint>

 <!--left rear 2nd joint-->
  <joint type="revolute" name="hinge_lr2">
    <parent link="base_left_rear"/>  
    <child link="link_left_rear_upper"/>
    <origin xyz="0 0 0" rpy="1.5707 0 0" />
    <axis xyz="0 1 0" rpy="0 0 0"/>
    <limit effort="30" velocity="1.0" lower="-2.2" upper="0.7" />
  </joint>
  
  <!--right fore 2nd joint--> 
  <joint type="revolute" name="hinge_rf2">
    <parent link="base_right_fore"/>  
    <child link="link_right_fore_upper"/>
    <origin xyz="0 0 0" rpy="1.5707 0 0" />
    <axis xyz="0 1 0" rpy="0 0 0"/>
    <limit effort="30" velocity="1.0" lower="-2.2" upper="0.7" />
  </joint>

  <!--right rear 2nd joint--> 
  <joint type="revolute" name="hinge_rr2">
    <parent link="base_right_rear"/>  
    <child link="link_right_rear_upper"/>
    <origin xyz="0 0 0" rpy="1.5707 0 0" />
    <axis xyz="0 1 0" rpy="0 0 0"/>
    <limit effort="30" velocity="1.0" lower="-2.2" upper="0.7" />
  </joint>
  
  <!--left fore 3rd joint--> 
  <joint type="revolute" name="hinge_lf3">
    <parent link="link_left_fore_upper"/>  
    <child link="link_left_fore_lower"/>
    <origin xyz="0 0 0.30" rpy="0 0 0" />
    <axis xyz="0 1 0" rpy="0 0 0"/>
    <limit effort="30" velocity="1.0" lower="-2.2" upper="0.7" />
  </joint>

  <!--left rear 3rd joint--> 
  <joint type="revolute" name="hinge_lr3">
    <parent link="link_left_rear_upper"/>  
    <child link="link_left_rear_lower"/>
    <origin xyz="0 0 0.30" rpy="0 0 0" />
    <axis xyz="0 1 0" rpy="0 0 0"/>
    <limit effort="30" velocity="1.0" lower="-2.2" upper="0.7" />
  </joint>
  
  <!--right fore 3rd joint--> 
  <joint type="revolute" name="hinge_rf3">
    <parent link="link_right_fore_upper"/>  
    <child link="link_right_fore_lower"/>
    <origin xyz="0 0 0.30" rpy="0 0 0" />
    <axis xyz="0 1 0" rpy="0 0 0"/>
    <limit effort="30" velocity="1.0" lower="-2.2" upper="0.7" />
  </joint>
 
  <!--right rear 3rd joint--> 
  <joint type="revolute" name="hinge_rr3">
    <parent link="link_right_rear_upper"/>  
    <child link="link_right_rear_lower"/>
    <origin xyz="0 0 0.30" rpy="0 0 0" />
    <axis xyz="0 1 0" rpy="0 0 0"/>
    <limit effort="30" velocity="1.0" lower="-2.2" upper="0.7" />
  </joint>

  <plugin name="gazebo_legged_robot_plugin" filename="libgazebo_legged_robot_plugin.so" >
    <hinge_lf1>hinge_lf1</hinge_lf1>
    <hinge_lf2>hinge_lf2</hinge_lf2>
    <hinge_lf3>hinge_lf3</hinge_lf3>
    <hinge_lr1>hinge_lr1</hinge_lr1>
    <hinge_lr2>hinge_lr2</hinge_lr2>
    <hinge_lr3>hinge_lr3</hinge_lr3>
    <hinge_rf1>hinge_rf1</hinge_rf1>
    <hinge_rf2>hinge_rf2</hinge_rf2>
    <hinge_rf3>hinge_rf3</hinge_rf3>
    <hinge_rr1>hinge_rr1</hinge_rr1>
    <hinge_rr2>hinge_rr2</hinge_rr2>
    <hinge_rr3>hinge_rr3</hinge_rr3>
    <gain>0.1 </gain>
    <sensor>sensor</sensor>
    <ray_sensor>laser</ray_sensor>
  </plugin>
</robot>
